Replace RPi.GPIO with rpi-lgpio for V3 future3

s-martin opened the pull request:
Following the discussion in https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/2313 replacing RPi.GPIO with [rpi-lgpio](https://rpi-lgpio.readthedocs.io/en/latest/index.html)

- [x] Clarify if we need a check mechanism to not accidentally install RPi.GPIO as dependency in the future - see comment from @AlvinSchiller https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/2469#issuecomment-2564880847
- [x] Make sure that rc522 does not install RPi.GPIO as dependency (—no-deps) - pointer: https://github.com/MiczFlor/RPi-Jukebox-RFID/blob/0ee5bc3f0da45ad935e7b046c4ce7653c9ed00a6/src/jukebox/components/rfid/configure/__init__.py#L29-L37
- [x] build rpi-lgpio also manually on Bullseye - see #2469 

@hoffie @varac

s-martin committed:
replace RPi.GPIO with rpi-lgpio

s-martin commented:
Anybody using a RC522 reader is welcome to test the change (also #2469 if you use V2) as I don't have a RC522 reader.

pabera commented:
Is the API interchangeable? It seems easy, but should be tested with some default components

pabera reviewed (commented):
It's a good idea, but I don't think we need it. Our installation does not qualify for reinstallation anyways. It should not be installed by default.

pabera commented on installation/routines/setup_jukebox_core.sh:
```diff
@@ -30,6 +30,8 @@ _jukebox_core_install_python_requirements() {
   source "$VIRTUAL_ENV/bin/activate"
 
   pip install --upgrade pip
+  # Remove RPi.GPIO, if still installed - see https://github.com/MiczFlor/RPi-Jukebox-RFID/issues/2313
```
It's a good idea, but I don't think we need it. Our installation does not qualify for reinstallation anyways. It should not be installed by default.

s-martin commented:
> * [ ]  Clarify if we need a check mechanism to not accidentally install RPi.GPIO as dependency in the future - see comment from @AlvinSchiller [Replace RPi.GPIO with rpi-lgpio for V2 legacy #2469 (comment)](https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/2469#issuecomment-2564880847)
> 

Just thought about checking the installed package (and making sure RPi.GPIO is not installed) at least in the CI, but I realized that we don’t have these CI checks in future3 yet.

s-martin commented:
I think for future3 to fix the rc522 issues https://github.com/hoffie/pi-rc522-gpiozero would be a good solution as we use gpiozero anyway in future3.

@pabera @AlvinSchiller

s-martin commented:
Added task to build rpi-lgpio manually on Bullseye - some users already tried to install on Bullseye (https://github.com/MiczFlor/RPi-Jukebox-RFID/discussions/2481#discussioncomment-11809284)

AlvinSchiller committed:
fix: update setup for RPi.GPIO replacement

AlvinSchiller committed:
fix: update RFIDReader pi-rc522 Lib for gpiozero support. remove RPi.GPIO references

AlvinSchiller committed:
Merge branch 'future3/develop' into future3/test/fix-rpi-gpio-bookworm

AlvinSchiller committed:
fix: pin "pi-rc522-gpiozero" lib to specific commit

AlvinSchiller committed:
fix: remove RPi.GPIO usage. Check for command existance and ignore if not available

AlvinSchiller committed:
fix: warning on gcc build

"format ‘%ld’ expects argument of type ‘long int’, but argument 2 has type ‘size_t’ {aka ‘unsigned int’}"

AlvinSchiller committed:
fix: update docs

AlvinSchiller committed:
fix: update comment

AlvinSchiller commented:
- updated rpi-lgpio installation (incl build lgpio for bullseye)
- add installation check that `RPi.GPIO` is not installed (after core dependencies)
- switched `pi-rc522` lib to `pi-rc522-gpiozero` fork from [hoffie](https://github.com/hoffie/pi-rc522-gpiozero)
- removed all direct RPi.GPIO references to be fully gpiozero compatible (e.g. usage of LgpioFactory)
  - formerly rpi specific host functions now check if command is available on system (tvservice for `hdmi_power_down` also not available on bookworm lite).

s-martin commented on installation/routines/setup_jukebox_core.sh:
```diff
@@ -24,13 +24,33 @@ _jukebox_core_install_os_dependencies() {
 
   pip install --upgrade pip
-  pip install --no-cache-dir -r "${INSTALLATION_PATH}/requirements.txt"
+  # Remove excluded libs, if installed - see - see https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/2470
```
```suggestion
  # Remove excluded libs, if installed - see https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/2470
```

s-martin commented on src/cli_client/pbc.c:
The open PR #2452 removes the C tool and we use just the Python tool.

AlvinSchiller replied:
Ahhh there it was. I had in mind this changed somehow but didn't find where 😄

I will revert this change to not cause merge conflicts.

AlvinSchiller committed:
Revert "fix: warning on gcc build"

This reverts commit 0be5cdb0e1b6f65ecb690cf01b83c492202b22cf.

AlvinSchiller committed:
fix: remove obsolete python setup.

pabera commented on installation/routines/setup_jukebox_core.sh:
```diff
@@ -24,13 +24,33 @@ _jukebox_core_install_os_dependencies() {
+    unzip ${lg_zip_filename} || exit_on_error
+    cd "${lg_filename}" || exit_on_error
+    make && sudo make install
+    cd "${INSTALLATION_PATH}" && sudo rm -rf "${tmp_path}"
```
Consider splitting the lgpio build logic into a separate function for better maintainability

pabera commented on requirements-excluded.txt:
```diff
@@ -0,0 +1,5 @@
+# Remove excluded libs, if installed - see https://github.com/MiczFlor/RPi-Jukebox-RFID/pull/2470
+# Libraries which must be excluded. These must be removed with
+# You need to uninstall these with `python -m pip uninstall -y -r requirements.txt`
```
Good documentation of purpose, but can we add more details about why RPi.GPIO needs to be excluded.

pabera commented on src/jukebox/components/rfid/hardware/rc522_spi/rc522_spi.py:
```diff
@@ -87,7 +86,7 @@ def __init__(self, reader_cfg_key):
                                    pin_irq=pin_irq,
                                    antenna_gain=antenna_gain,
-                                   pin_mode=GPIO.BCM)
+                                   pin_mode='BCM')
```
Does this accept strings?

AlvinSchiller replied:
Yep, [see](https://github.com/hoffie/pi-rc522-gpiozero/blob/1dc878c8a3d6f32ad6601877b00e7316d72811d1/pirc522/rfid.py#L13).
 It would work also with `GPIO.BCM` but i wanted to eliminate the import and give the possibility to use other gpio libs for gpiozero.

AlvinSchiller committed:
fix: update description for excluded libs

AlvinSchiller committed:
fix: moved lgpio build into function. fixed needed packages. use download function

AlvinSchiller committed:
fix: enabled output for wget.
